core/src/main/scala/higherkindness/mu/rpc/srcgen/avro/AvroSrcGenerator.scala

```diff
@@ -152,40 +138,65 @@ final case class AvroSrcGenerator(
            )
          } else Nil)
 
-    val serviceParams: String = (serializationType.toString +: extraParams).mkString(",")
+    val serviceParams = (serializationType.toString +: extraParams).mkString(",")
+
+    val requestLines = {
+      val result = protocol.getMessages.asScala.toList.traverse {
+        case (name, message) =>
+          val comment = Seq(Option(message.getDoc).map(doc => s"  /** $doc */")).flatten
+          buildMethodSignature(name, message.getRequest, message.getResponse).map { content =>
+            comment ++ Seq(content, "")
+          }
+      }
+      result.map(_.flatten)
+    }
 
-    val serviceLines =
-      if (requestLines.isEmpty) Seq.empty
-      else
-        Seq(
-          s"@service($serviceParams) trait ${protocol.getName}[F[_]] {",
-          ""
-        ) ++ requestLines :+ "}"
+    val outputCode = requestLines.map { requests =>
+      val serviceLines =
+        if (requests.isEmpty) Seq.empty
+        else {
+          Seq(
+            s"@service(${serviceParams}) trait ${protocol.getName}[F[_]] {",
+            ""
+          ) ++ requests :+ "}"
+        }
+      packageLines ++ importLines ++ messageLines ++ serviceLines
+    }
 
-    outputPath -> (packageLines ++ importLines ++ messageLines ++ serviceLines)
+    outputPath -> outputCode
   }
 
-  private def parseMessage(name: String, request: Schema, response: Schema): String = {
-    val args = request.getFields.asScala
-    if (args.size > 1)
-      throw ParseException("RPC method has more than 1 request parameter")
-    val requestParam = {
-      if (args.isEmpty)
-        s"$DefaultRequestParamName: $EmptyType"
-      else {
-        val arg = args.head
-        if (arg.schema.getType != Schema.Type.RECORD)
-          throw ParseException("RPC method request parameter is not a record type")
-        s"${arg.name}: ${arg.schema.getFullName}"
-      }
-    }
-    val responseParam = {
-      if (response.getType == Schema.Type.NULL) EmptyType
-      else s"${response.getNamespace}.${response.getName}"
+  def validateRequest(request: Schema): ErrorsOr[String] = {
+    val requestArgs = request.getFields.asScala
+    if (requestArgs.size > 1)
+      (s"RPC method ${request.getName} has more than 1 request parameter").invalidNel
```

Comment:
This line is dead code, isn't it? Not sure why it didn't trigger a compiler warning.

I think a pattern match would be clearer than nested if's, and would avoid sneaky dead code issues like this one:

```scala
requestArgs.toList match {
  case Nil =>
    s"$DefaultRequestParamName: $EmptyType".validNel
  case x :: Nil if x.schema.getType == Schema.Type.RECORD =>
    s"${requestArg.name}: ${requestArg.schema.getFullName}".validNel
  case x :: Nil =>
    s"RPC method request parameter '${requestArg.name}' has non-record request type '${requestArg.schema.getType}'".invalidNel
  case _ =>
    s"RPC method ${request.getName} has more than 1 request parameter".invalidNel
}
```